MaxText/input_pipeline/_hf_data_processing.py

```diff
@@ -32,6 +32,116 @@
+
+  dataset = dataset.select_columns(text_columns + [image_column])
+  dataset = dataset.map(
+      _input_pipeline_utils.reformat_prompt,
+      fn_kwargs={"column": text_columns[0], "model_name": config.model_name},
+  )
+  dataset = dataset.map(
+      _input_pipeline_utils.reformat_response,
+      fn_kwargs={"column": text_columns[1], "model_name": config.model_name},
+  )
+  if image_column != "images":
+    dataset = dataset.rename_column(image_column, "images")
+
+  dataset = dataset.map(
+      _input_pipeline_utils.pre_process_image_sft,
+      fn_kwargs={"image_column": "images", "model_name": config.model_name},
+  )
+
+  tokenizer = transformers.AutoTokenizer.from_pretrained(
+      config.tokenizer_path,
+      add_bos_token=False,
+      add_eos_token=False,
+      legacy=False,
+      token=config.hf_access_token,
+  )
+  if tokenizer.pad_token_id is not None:
+    pad_id = tokenizer.pad_token_id
+  elif tokenizer.unk_token_id is not None:
+    pad_id = tokenizer.unk_token_id
+  else:
+    pad_id = -1
+
+  dataset = dataset.map(
+      _input_pipeline_utils.tokenization,
+      batched=True,
+      fn_kwargs={
+          "hf_tokenizer": tokenizer,
+          "truncation": False,
+          "max_length": config.max_target_length,
+          "column_names": text_columns,
+      },
+  )
+  dataset = dataset.map(
+      _input_pipeline_utils.prepare_text_for_image_fusion,
+      fn_kwargs={"column_name": text_columns[0], "model_name": config.model_name},
+  )
+
+  dataset = _input_pipeline_utils.HFDataSource(
+      dataset=dataset,
+      dataloading_host_index=dataloading_host_index,
+      dataloading_host_count=dataloading_host_count,
+      num_threads=1,
+      generate_padding_example=True,
+      max_target_length=config.max_target_length,
+      data_column_names=text_columns,
+  )
+  operations = []
+  operations.append(
+      _input_pipeline_utils.SFTPromptMaskingVision(
+          query_column=text_columns[0],
+          response_column=text_columns[1],
+          max_target_length=config.max_target_length,
+          unk_id=pad_id,
+      )
+  )
+  # TODO(aireenmei, hengtaoguo): support packing
+  operations.append(_input_pipeline_utils.PadToMaxLength(config.max_target_length, pad_id))
+  operations.append(grain.Batch(batch_size=global_batch_size // jax.process_count(), drop_remainder=True))
+  operations.append(_input_pipeline_utils.ShiftData(ignored_ids=[pad_id], axis=1))
+  dummy_index_sampler = grain.IndexSampler(
+      num_records=len(dataset),
+      num_epochs=1,
+      shard_options=grain.ShardOptions(
+          shard_index=dataloading_host_index, shard_count=dataloading_host_count, drop_remainder=False
+      ),
+      shuffle=False,
+      seed=0,
+  )
+
+  dataloader = grain.DataLoader(
+      data_source=dataset,
+      operations=operations,
+      sampler=dummy_index_sampler,
+      worker_count=1,  # only supports <=1 for now, more workers results in duplicated data
```

Comment:
why worker_count=1 results duplicate data? is it data bound with `worker_count=1` ?